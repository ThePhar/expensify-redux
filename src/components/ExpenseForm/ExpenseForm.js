import moment               from 'moment';
import React                from 'react';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };

    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please provide description and amount.' }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        // required
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    /><br />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                        // required
                    /><br />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        id="date_picker"
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    /><br />
                    <textarea
                        placeholder="Add a note for your expense."
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    /><br />
                    <button>Submit</button>
                </form>
            </div>
        );
    }
}
