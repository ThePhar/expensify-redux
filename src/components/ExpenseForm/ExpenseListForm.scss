.form-layout {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100vw;

    form {
        width: 100%;
        padding: 0 1rem;
    }

    input {
        border: 1px solid #dbdbdb;
        font-weight: 200;
        font-size: 19px;
        line-height: 24px;
        color: #484848;
        padding: 11px;
        height: 48px;
        width: 100%;
        margin-bottom: 0.5rem;
    }

    textarea {
        border: 1px solid #dbdbdb;
        font-weight: 200;
        font-size: 19px;
        line-height: 24px;
        color: #484848;
        padding: 11px;
        height: calc(48px * 4);
        margin-bottom: 0.5rem;
        width: 100%;
    }

    .SingleDatePickerInput__withBorder {
        border: none;
    }
}