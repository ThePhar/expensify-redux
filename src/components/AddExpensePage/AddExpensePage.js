import React from "react";
import ExpenseForm from "../ExpenseForm/ExpenseForm";

const AddExpensePage = () => (
    <div>
        <h1>Add Expense</h1>
        <ExpenseForm />
    </div>
);

export default AddExpensePage;