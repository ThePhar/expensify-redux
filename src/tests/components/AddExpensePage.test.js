import { shallow }        from 'enzyme';
import React              from 'react';
import { AddExpensePage } from '../../components/AddExpensePage/AddExpensePage';
import ExpenseForm        from '../../components/ExpenseForm/ExpenseForm';
import expenses           from '../fixtures/expenses';

let startAddExpense, history, wrapper;
beforeEach(() => {
    startAddExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<AddExpensePage startAddExpense={startAddExpense} history={history} />);
});

test('Should render AddExpensePage correctly.', () => {
    expect(wrapper).toMatchSnapshot();
});

test('Should handle onSubmit', () => {
    wrapper.find(ExpenseForm).prop('onSubmit')(expenses[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startAddExpense).toHaveBeenLastCalledWith(expenses[1]);
});