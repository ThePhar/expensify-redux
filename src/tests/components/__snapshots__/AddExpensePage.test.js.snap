// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render AddExpensePage correctly. 1`] = `
<div>
  <h1
    className="ma4"
  >
    Add A New Expense
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
